android/src/main/java/com/robinpowered/Frequency/FrequencyModule.java

```diff
@@ -0,0 +1,141 @@
+    private static final String MODULE_NAME = "Frequency";
+    private static final int SAMPLE_RATE = 44100;
+
+    private AudioSession audioSession;
+
+    public FrequencyModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    @Override
+    public String getName() {
+        return MODULE_NAME;
+    }
+
+    private void cancel() {
+        audioSession.cancel();
+        audioSession = null;
+    }
+
+    private void complete() {
+        audioSession.complete();
+        audioSession = null;
+    }
+
+    @ReactMethod
+    public void playFrequency(double frequency, double duration, final Promise promise) {
+        if (audioSession != null && audioSession.isPlaying()) {
+            cancel();
+        }
+
+        final int dur = (int) duration;
+        final int numOfSamples = dur * SAMPLE_RATE;
+
+        final double sample[] = new double[numOfSamples];
+        final byte soundData[] = new byte[2 * numOfSamples];
+
+        for (int i = 0; i < numOfSamples; ++i) {
+            sample[i] = Math.sin(2 * Math.PI * i / (SAMPLE_RATE/frequency));
+        }
+
+        // convert to 16 bit pcm sound array
+        // assumes the sample buffer is normalised.
+        int idx = 0;
+        for (double dVal : sample) {
+            short val = (short) (dVal * 32767);
+            soundData[idx++] = (byte) (val & 0x00ff);
+            soundData[idx++] = (byte) ((val & 0xff00) >>> 8);
+        }
+
+        AudioTrack track;
+
+        // create audio track - methods used differ based on OS version
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+            track = new AudioTrack(AudioManager.STREAM_MUSIC, SAMPLE_RATE, AudioFormat.CHANNEL_OUT_MONO,
+                    AudioFormat.ENCODING_PCM_16BIT, numOfSamples, AudioTrack.MODE_STATIC);
+        } else {
+            track = new AudioTrack.Builder()
+                    .setAudioFormat(new AudioFormat.Builder()
+                            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
+                            .setSampleRate(SAMPLE_RATE)
+                            .setChannelMask(AudioFormat.CHANNEL_OUT_MONO).build())
+                    .setBufferSizeInBytes(numOfSamples).build();
+        }
+
+        // push sample into AudioTrack object
+        track.write(soundData, 0, numOfSamples);
+
+        // callback when track finishes playing
+        track.setNotificationMarkerPosition(numOfSamples / 2);
+
+        track.setPlaybackPositionUpdateListener(new AudioTrack.OnPlaybackPositionUpdateListener() {
+            @Override
+            public void onPeriodicNotification(AudioTrack track) {
+                // no-op
+            }
+
+            @Override
+            public void onMarkerReached(AudioTrack track) {
+                FrequencyModule.this.complete();
+            }
+        });
+
+        audioSession = new AudioSession(track, promise);
+
+        // play track
+        track.play();
+    }
+}
+
+class AudioSession {
```

Comment:
Let's on merging - I've got a few thoughts on this class structure. 
Will discuss off of github, then we can circle back and document what we come up with.

Reply:
Responded in https://github.com/robinpowered/react-native-frequency/pull/1#discussion_r160264331

I don't really agree with using inheritance here for those reasons. While they both accomplish the same outcome, a subclass has a different semantic meaning. We're not creating a different type of audio track, we're aiming to couple a track with a promise -- one does not own the other, they are siblings in a API transaction from the client, which is what brought light to the AudioSession instance that manages the siblings

Reply:
I think this section of composition over inheretence describes the design pattern well: https://en.wikipedia.org/wiki/Composition_over_inheritance#Benefits

>  It is more natural to build business-domain classes out of various components than trying to find commonality between them and creating a family tree.

In this module, we're not trying to create a new classification of audio tracks, we're aiming to couple the promise and track for consistency and controlled access